### PR TITLE
feat(orchestrator): support pagination for /instances

### PR DESCRIPTION
**Pull request by JudeNiroshan**

### CODE REVIEW

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 1]
```diff
@@ -100,10 +103,19 @@ export class DataIndexService {
...
   }
 
-  public async fetchProcessInstances(): Promise<ProcessInstance[] | undefined> {
-    const graphQlQuery =
-      '{ ProcessInstances ( orderBy: { start: ASC }, where: {processId: {isNull: false} } ) { id, processName, processId, businessKey, state, start, end, nodes { id }, variables, parentProcessInstance {id, processName, businessKey} } }';
-
+  public async fetchProcessInstances(
+    pagination: Pagination,
+  ): Promise<ProcessInstance[] | undefined> {
+    pagination.sortField = FETCH_PROCESS_INSTANCES_SORT_FIELD;
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: de04daf47e4d92f4c56e5fc4b206779a08f8a327

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 2]
```diff
@@ -100,10 +103,19 @@ export class DataIndexService {
...
+
+    const graphQlQuery = buildGraphQlQuery({
+      type: 'ProcessInstances',
+      queryBody:
+        'id, processName, processId, state, start, lastUpdate, end, nodes { id }, variables, parentProcessInstance {id, processName, businessKey}',
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: c762027b95531acf63461e3c9738aba476eb71c3

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 3]
```diff
@@ -214,8 +233,13 @@ export class DataIndexService {
   public async fetchProcessInstance(
     instanceId: string,
   ): Promise<ProcessInstance | undefined> {
-    const graphQlQuery = `{ ProcessInstances (where: { id: {equal: "${instanceId}" } } ) { id, processName, processId, serviceUrl, businessKey, state, start, end, nodes { id, nodeId, definitionId, type, name, enter, exit }, variables, parentProcessInstance {id, processName, businessKey}, error { nodeDefinitionId, message} } }`;
-
+    const graphQlQuery = buildGraphQlQuery({
+      type: 'ProcessInstances',
+      queryBody:
+        'id, processName, processId, state, start, lastUpdate, end, nodes { id, nodeId, definitionId, type, name, enter, exit }, variables, parentProcessInstance {id, processName, businessKey}, error { nodeDefinitionId, message}',
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: c762027b95531acf63461e3c9738aba476eb71c3 0f88c42611051bf77cbec4f66b4f53991eabdd07 f8836ff22150b240a0ffb78f6da1f6a97b1468d8

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 4]
```diff
@@ -179,9 +191,16 @@ export class DataIndexService {
...
   public async fetchProcessInstanceJobs(
     instanceId: string,
+    pagination: Pagination,
   ): Promise<Job[] | undefined> {
-    const graphQlQuery = `{ Jobs (where: { processInstanceId: { equal: "${instanceId}" } }) { id, processId, processInstanceId, rootProcessId, status, expirationTime, priority, callbackEndpoint, repeatInterval, repeatLimit, scheduledId, retries, endpoint, nodeInstanceId, executionCounter } }`;
-
+    const graphQlQuery = buildGraphQlQuery({
+      type: 'Jobs',
+      queryBody:
+        'id, processId, processInstanceId, rootProcessId, status, expirationTime, priority, callbackEndpoint, repeatInterval, repeatLimit, scheduledId, retries, lastUpdate, endpoint, nodeInstanceId, executionCounter',
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: f8836ff22150b240a0ffb78f6da1f6a97b1468d8

File: plugins/orchestrator-backend/src/service/DataIndexService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: f8836ff22150b240a0ffb78f6da1f6a97b1468d8

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 5]
```diff
@@ -175,11 +187,15 @@ export class DataIndexService {
...
-    limit: number,
-    offset: number,
+    pagination: Pagination,
   ): Promise<ProcessInstance[]> {
-    const graphQlQuery = `{ ProcessInstances(where: {processId: {equal: "${workflowId}" } }, pagination: {limit: ${limit}, offset: ${offset}}) { processName, state, start, end } }`;
-
+    const graphQlQuery = buildGraphQlQuery({
+      type: 'ProcessInstances',
+      queryBody: 'processName, state, start, end',
+      whereClause: `processId: {equal: "${workflowId}"`,
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: 420fa3c

File: plugins/orchestrator-backend/src/service/SonataFlowService.ts [hunk 6]
```diff
@@ -442,7 +444,8 @@ export class SonataFlowService {
           counter++;
         }
       }
-      offset += limit;
+      if (pagination.offset && pagination.limit)
+        pagination.offset += pagination.limit;
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/SonataFlowService.ts]: [8fa7dca](https://github.com/janus-idp/backstage-plugins/pull/1251/commits/8fa7dca15d704f9cffe60a955067ef8383d22583)

**Comment by caponetto** [comment]
@JudeNiroshan Please resolve the conflicts so we can merge this PR.

File: plugins/orchestrator-backend/src/service/DataIndexService.ts [hunk 7]
```diff
@@ -91,25 +98,13 @@ export class DataIndexService {
...
+      type: 'ProcessDefinitions',
+      queryBody: 'id, name, version, type, endpoint, serviceUrl',
+    });
+    this.logger.debug(`GraphQL query: ${graphQlQuery}`);
     this.logger.info(`getWorkflowDefinitions() called:  ${this.dataIndexUrl}`);
-    const result = await this.client.query(QUERY, {});
-
-    this.logger.debug(
-      `Get workflow definitions result: ${JSON.stringify(result)}`,
-    );
```
Comment (caponetto) [inline plugins/orchestrator-backend/src/service/DataIndexService.ts:112]: You have removed this log probably during conflict resolution. Please put it back.

Comment (JudeNiroshan) [reply plugins/orchestrator-backend/src/service/DataIndexService.ts]: [fb3b435](https://github.com/janus-idp/backstage-plugins/pull/1251/commits/fb3b435c53c7d1cfbb5ae51018affccdeede9c2d)

**Comment by caponetto** [comment]
@JudeNiroshan I tested with ~30 instances but the table only shows 10 and doesn't show the pagination options.
Could you please check? `main` works fine.

**Comment by JudeNiroshan** [comment]
Closing this PR as the requirements have changed. The new PR is here -> https://github.com/janus-idp/backstage-plugins/pull/1313

**Pull request closed without merging**